docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,281 @@
+# MeshLoadBalancingStrategy - locality aware
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/2689
+
+## Context and Problem Statement
+Right now we have really simple locality aware load balancing, we only care about zones. Dataplane will prefer sending traffic to local zone with fallback to other zones. We would like to add more sophisticated mechanism of locality aware load balancing. We would like to allow user to specify load balancing rules based on tags. For example, big cluster is deployed in multiple AZ’s, and we would like to prefer staying in a single AZ to reduce traffic cost.
+Considered Options
+
+## Decision Outcome
+
+## Solution
+As stated earlier, we would like to improve our current locality aware load balancing, and make it more flexible for users.
+### Use cases
+After gathering client's expectations and a couple of brainstorming sessions, we ended up with multiple use cases for locality aware load balancing:
+
+1. Keeping traffic in single zone
+2. Load balancing based on priorities within a zone. Example: node → AZ → region
+3. Prefer traffic in single zone with fallback to other zones
+4. Load balance equally to list of zones
+5. When leaving zone, target specific zones first. Example zones physically close to each other
+6. Load balance equally to local dpp and other zone
+7. Exclude zone from load balancing
+
+### API design
+Looking at use cases, it is clear that we have two types of load balancing strategies, inside zone and cross zone. We can easily route to services with proper tags inside the zone, but immediately after we cross zone border and hit Zone Ingress we lose the ability to control locality.
+Because of this we have decided that we will split our API in two sections: `inZone` and `crossZone`.
+
+Here is the full example of policy:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: ["k8s.io/node", "k8s.io/az"] # (1)
+      crossZone: # (2)
+        rules: # (3)
+          - groups: # (4)
+            - ["zone-1", "zone-2"]
+            - ["zone-3", "zone-4"]
+          - groups:
+            - ["zone-5"]
+        fallback: any # (5)
+```
+
+
+(1) In `inZone` section, you configure the priorities of tags used for load balancing.
+
+(2) In `crossZone` section, you configure zones priority.
+
+(3) In `rules` section, you configure list of zone groups priorities, first groups will have the highest load balancing priority and last in order will have the lowest priority.
+
+(4) In `groups` section, you configure logical zone groups. Let's assume `zone-1` and `zone-2` are located in the same datacenter, so you want to load balance equally between them. `Zone-5` is located in another datacenter so it should have lower priority, that is why it is in the second groups entry.
```

Comment:
Little confused. How do I get routed if I'm a DPP in zone-5?

Reply:
Also with the first set of groups do I route to any zone regardless of which zone I'm part of or does it still favour self?

Reply:
What happens when the group is empty?

Reply:
So the inzone traffic has the highest priority. In this example we would have 
Priority 0: the same k8s.io/node
Priority 1: the same k8s.io/az
Priority 2: Zone 1. zone-2
Priority 3: Zone5


If you don't define any crossZone we want to assume that you want to keep traffic within the zone but fallback to any:
```
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"]
      crossZone: {}
```

When groups are empty we assume you want to treat all zones the same.

```
  defaults:
    localityAwareness:
      inZone: ["k8s.io/node", "k8s.io/az"]
      crossZone:
       rules:
         - groups: []
       
```

Does it make sense?

Reply:
this example is no longer valid